### PR TITLE
handles project deletion

### LINKED ISSUE #1072: Handle Project deletion
Add finalizer to a project instance
Remove corresponding `ProjectGroup` from the `User` object
...
Remove finalizer
child of #966

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: handles project deletion, in particular we will remove references to the project from the corresponding users objects. Additionally all dependant resources - the ones that have `OwnerReferences` set will be automatically GC`ed

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1072

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/provider/kubernetes/project.go [hunk 1]
```diff
@@ -60,8 +62,9 @@ func (p *ProjectProvider) New(user *kubermaticapiv1.User, projectName string) (*
 		ObjectMeta: metav1.ObjectMeta{
 			OwnerReferences: []metav1.OwnerReference{
 				{
-					APIVersion: user.APIVersion,
-					Kind:       user.Kind,
+					// don't know why but sometimes user.Kind and user.APIVersion are empty
```
Comment (p0lyn0mial) [inline api/pkg/provider/kubernetes/project.go:71]: this is weird...

File: api/pkg/controller/rbac/sync.go [hunk 2]
```diff
@@ -2,12 +2,22 @@ package rbac
 
 import (
 	"fmt"
+
 	"github.com/golang/glog"
 
 	kubermaticv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
 
+	"strings"
+
```
Comment (mrIncompetent) [inline api/pkg/controller/rbac/sync.go]: Could you group the go-sdk imports?

File: api/pkg/controller/rbac/sync.go [hunk 3]
```diff
@@ -112,3 +179,46 @@ func (c *Controller) ensureProjectRBACRoleBinding(project *kubermaticv1.Project)
...
+			if strings.HasSuffix(pg.Group, project.Name) {
+				continue
+			}
+			updatedProjectGroup = append(updatedProjectGroup, pg)
+		}
+		if len(updatedProjectGroup) != len(sharedUser.Spec.Projects) {
+			user := sharedUser.DeepCopy()
+			user.Spec.Projects = updatedProjectGroup
+			_, err = c.kubermaticClient.KubermaticV1().Users().Update(user)
+			if err != nil {
```
Comment (mrIncompetent) [inline api/pkg/controller/rbac/sync.go]: Inline:
```go
if _, err := c.kubermaticClient.KubermaticV1().Users().Update(user); err != nil {
  return err
}
```

File: api/pkg/controller/rbac/sync.go (hunk 2 above)
Comment (p0lyn0mial) [reply api/pkg/controller/rbac/sync.go]: sure, np.

File: api/pkg/controller/rbac/sync.go (hunk 3 above)
Comment (p0lyn0mial) [reply api/pkg/controller/rbac/sync.go]: ok.

File: api/pkg/provider/kubernetes/project.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/provider/kubernetes/project.go:71]: No idea why this happens. Maybe that's why `metav1.NewControllerRef` takes a `GroupVersionKind` as second parameter

Comment (p0lyn0mial) [reply api/pkg/provider/kubernetes/project.go:71]: I intentionally left reading `project.APIVersion` and `project.Kind` to see if this is only specific to user

https://github.com/kubermatic/kubermatic/blob/5582a861e8f90ea77ac035cdb03e52f3eba89d88/api/pkg/controller/rbac/sync.go#L116

**Comment by p0lyn0mial** [comment]
comments addressed.

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**